Chatkit/Chatkit/Chatkit.swift

```diff
@@ -99,102 +101,77 @@ public class Chatkit {
     }
     
-    /// Creates an instance of `AvailableRoomsProvider`.
+    /// Creates an instance of `AvailableRoomsRepository`.
     ///
     /// - Parameters:
-    ///     - completionHandler: A completion handler which will be called when the `AvailableRoomsProvider` is ready, or an `Error` occurs creating it.
-    public func createAvailableRoomsProvider(completionHandler: @escaping (AvailableRoomsProvider?, Error?) -> Void) {
+    ///     - completionHandler: A completion handler which will be called when the `AvailableRoomsRepository` is ready, or an `Error` occurs creating it.
+    public func createAvailableRoomsRepository(completionHandler: @escaping (AvailableRoomsRepository?, Error?) -> Void) {
         // TODO: Implement
         completionHandler(nil, nil)
     }
     
-    // MARK: - Constructing real time data providers
+    // MARK: - Constructing real time data repositories
     
-    /// Creates an instance of `JoinedRoomsProvider`.
+    /// Creates an instance of `JoinedRoomsRepository`.
     ///
     /// This will provide access to a real time set of `Room`s that the current user is a member of.
-    ///
-    /// - Parameters:
-    ///     - completionHandler: A completion handler which will be called when the `JoinedRoomsProvider` is ready, or an `Error` occurs creating it.
-    public func createJoinedRoomsProvider(completionHandler: @escaping (JoinedRoomsProvider?, Error?) -> Void) {
+    public func createJoinedRoomsRepository() -> JoinedRoomsRepository {
+        let filter = JoinedRoomsRepository.Filter()
+        let buffer = ConcreteBuffer(filter: filter, dependencies: self.dependencies)
+        let connectivityMonitor = ConcreteConnectivityMonitor(subscriptionType: .user, dependencies: self.dependencies)
         
-        switch connectionStatus {
-
-        case .connected:
-            
-            // TODO: Implement
-            let currentUser = User(identifier: "identifier",
-                                   name: "name",
-                                   avatar: nil,
-                                   presenceState: .online,
-                                   customData: nil,
-                                   createdAt: Date(),
-                                   updatedAt: Date())
-            
-            let joinedRoomsProvider = JoinedRoomsProvider(currentUser: currentUser, dependencies: dependencies)
-            
-            Self.execute(completionHandler, onMainThreadWith: joinedRoomsProvider, nil)
-        
-        case .disconnected:
-            let error = ChatkitError.disconnected
-            Self.execute(completionHandler, onMainThreadWith: nil, error)
-            
-        case .connecting:
-            let error = ChatkitError.connecting
-            Self.execute(completionHandler, onMainThreadWith: nil, error)
-        }
-
+        return JoinedRoomsRepository(buffer: buffer, connectivityMonitor: connectivityMonitor, dependencies: self.dependencies)
```

Comment:
There was one drawback I've noticed with the `Buffer` and `ConnectivityMonitor` interfaces in that they communicate via delegation AND also expose a `connectionState`/`state` property.  I think in an ideal world we would just have the delegation. 

I'm not 100% certain if its possible to solve this issue because we somehow need to get at the "initial" value.  I did wonder however whether the solution would be to perform the registration of the Store listeners here and then injecting the initial Store value in.  

Like I said I don't know if this is actually possible.   Do you have any thoughts on the interfaces or this idea?

Reply:
Let's leave this for now.  I have since realised that there is some outstanding work to do around ConnectivityMonitor and SubscriptionManager (that I mentioned yesterday) so we could potentially address this then.